Add support for HDR imagery to the PNG format (#12)
w3c/ColorWeb-CG: png_cicp_chunk_proposal.md

# Adding support for HDR imagery to the PNG format
Authors (alphabetical): Chris Blume, Chris Seeger, Pierre-Anthony Lemieux

Status: Draft

## Problem to be solved
The gAMA chunk of the Portable Network Graphics (PNG) format (specified in [PNG]) parameterized the transfer function of the image as a power law. As such, it cannot model the Reference PQ or HLG OTFs specified in [BT2100-1], which are commonly used for HDR images.

> This specification uses the existing iCCP chunk to unambiguously signal the color system of an image that uses the Reference PQ EOTF specified in [BT2100-1]. It also allows graceful processing by decoders that do not conform to this specification by recommending fallback values for the gAMA chunk, cHRM chunk, and embedded ICC profile.

Reviewer: "This specification" which specification?

> An existing W3C group note (https://www.w3.org/TR/png-hdr-pq/) specifies an approach which is limited: it supports signaling of only the ITU BT.2100 PQ EOTF and uses magic values in the iCCP chunk to signal color spaces.

Reviewer: I guess "magic values" is sufficient to indicate that this is a really bad idea. Specifically, the ICC profile, if applied, does not give the correct rendering and must be ignored :) and only its presence is used as a verbose way to label the content. This proposal is thus a significant improvement on that, in terms of internal consistency and also filesize.

> An existing W3C group note (https://www.w3.org/TR/png-hdr-pq/) specifies an approach which is limited: it supports signaling of only the ITU BT.2100 PQ EOTF and uses magic values in the iCCP chunk to signal color spaces.

Reviewer: Also, add that note to the informative references

[H.273](https://www.itu.int/rec/T-REC-H.273/en) specifies a controlled vocabulary for the parameterization of
color space information:
* Color Primaries
* Transfer Function

> * Matrix Coefficients (only required if essence is stored as Y’CbCr)

Reviewer: So, PNG does not support Y'CbCr; only RGB, RGBA, G, and GA.
Why is this Matrix Coefficients field needed?

Use [ITU-T Series H Supplement 19](https://www.itu.int/rec/T-REC-H.Sup19-201910-I) “recommendations” to signal the controlled vocabulary for image formats commonly used for baseband linear broadcasts or file-based Video-on-Demand(VOD) services.
* Tables within this document summarize values most commonly used in broadcast for 47 different standards documents including H.273

## Basic requirements
* does not break current implementations

> * extensible signaling of color space based on H.273

Reviewer 2: Why only H.273?   Why not support other forms of HDR?

Author: I don't know enough here to have answer.
Would you mind elaborating?

Reviewer 3: @lrosenthol [H.273](https://www.itu.int/rec/T-REC-H.273/en) supports common SDR/HDR formats. Do you know of any HDR format that PNG should support but is not listed in H.273?

> * does not require the presence of iCCP chunk and embedded ICC profiles

Reviewer 2: Why is this a requirement?

Reviewer: Filesize is one reason, a chunk with a 7 byte payload is smaller than an ICC profile with zlib compression.

Reviewer 2: True, but it doesn't give rich fallback then.  Also doesn't align with the work going on at the ICC to consider a `CICP` tag.

Author: I'm not sure I follow.
You are suggesting it should require the other chunk? Would you mind helping me understand why it should be required to have both?

Would it be an acceptable fallback to allow both chunks but not require both chunks? (If so, we probably want to explicitly specify how to handle a file with both chuns.)

Reviewer 3: Per the 2021-01-12 meeting:

>     - two challenges re: signaling colorspace in the ICC profile in contrast to signaling colorspace directly in the image:
>         - adds a level of indirection
>         - ICC profile can/is be easily stripped

I think HDR-in-PNG signaling should not be contingent on ICC profiles, although ICC profiles can be included, e.g. for fallback.

Reviewer 2: But now that means that PNG has its own model for colorspace specification that differs from other formats (eg. JPEG)

Reviewer 2: Also, what if the embedded ICC profile *has* the new tag for HDR - why not support that?

Reviewer: > 
> 
> But now that means that PNG has its own model for colorspace specification that differs from other formats (eg. JPEG)

PNG also has its own PNG-specific way to signal sRGB, that doesn't use an ICC profile.
EXIF has an EXIF-specific way to signal either sRGB or Adobe 1998 RGB, without an ICC profile

> * cICP chunk comes before frame data

Reviewer 2: Do you mean before the `IDAT` chunk?   Can you clarify this in terms of which chunk can come before and/or after (as per PNG spec, 5.6)

Reviewer: I agree with @lrosenthol that "before IDAT" is the best way to express this.

Author: Agreed. I will update it.

## Strawman approach

> Define a cICP chunk that contains the 7 bytes necessary to carry the

Reviewer 2: `cICP` would not be a valid chunk name (as per PNG spec, 5.4), specifically the four character since I would consider it "safe to copy" which means it should be lowercase - thus `cICp`

Reviewer: It is not clear to me whether this is intended to be safe to copy or not. But  iCCP  is also not safe to copy.

Reviewer: Nor indeed is `gAMA` and `cHRM`

Reviewer 2: Why would you consider the profile and gamma info not safe to copy?  They are absolutely safe to copy since the assumption is that you aren't changing the colorspace of the image data when processing it.

Author: Looking at http://w3.org/TR/PNG-Structure.html#Chunk-naming-conventions I think I understand why those chunks are unsafe to copy and why this new chunk should be as well.

"If the program has made _any_ changs to _critical_ chunks, ... then unrecognized unsafe chunks must **not** be copied to the output PNG file. (Of course, if the program **does** recognize the chunk, it can choose to output an appropriately modified version.)"

So for example, if the program doesn't understand the HDR chunk and modifies some pixel values, it might do so in the wrong color space. I don't think we want to preserve the incorrect color space. I think we want to throw out the HDR information and now have a file in sRGB, for example.

H.273 color space parameters:

* COLPRIMS, 2 bytes, One of the ColourPrimaries enumerated values specified in Rec. ITU-T H.273 | ISO/IEC 23001-8
* TRANSFC, 2 bytes, One of the TransferCharacteristics enumerated values specified in Rec. ITU-T H.273 | ISO/IEC 23001-8
* MATCOEFFS, 2 bytes, One of the MatrixCoefficients enumerated values specified in Rec. ITU-T H.273 | ISO/IEC 23001-8
* VIDFRNG, 1 byte, Value of the VideoFullRangeFlag specified in Rec. ITU-T H.273 | ISO/IEC 23001-8

> [ed.: these are inspired from recent JPEG standards that incorporate

Reviewer 2: JPEG-XL uses Rec. ITU-T H.273 | ISO/IEC 23091-2:2019(E) as the normative reference - why are you using 23001-8?

Reviewer 2: I also don't see how these four values map into the values described in JPEG-XL A.4 (Color Encoding).

Reviewer 3: Rec. ITU-T H.273 was developed as twin-text with ISO/IEC 23001-8, but it looks like ISO has replaced ISO/IEC 23001-8 with ISO/IEC 23091-2.

H.273 color space parameters]

> 

Reviewer: Yes, priority in the presence of conflicting information is essential to specify.

As an example, look at what the PNG spec [says about the `iCCP` chunk in the presence of `gAMA` and cHRM` for readers and writers](https://www.w3.org/TR/PNG/#11iCCP).

An existing W3C group note, [BT2100-in-PNG]  specifies an approach which is limited: it supports signaling of only the ITU BT.2100 PQ EOTF and uses magic values in the iCCP chunk to signal color spaces.

## Basic requirements
* does not break current implementations
* extensible signaling of color space based on H.273
* does not require the presence of iCCP chunk and embedded ICC profiles
* cICP chunk comes before IDAT chunk

## Strawman approach
Define a cICP chunk that contains the 7 bytes necessary to carry the
H.273 color space parameters:

* COLPRIMS, 2 bytes, One of the ColourPrimaries enumerated values specified in Rec. ITU-T H.273 | ISO/IEC 23091-2
* TRANSFC, 2 bytes, One of the TransferCharacteristics enumerated values specified in Rec. ITU-T H.273 | ISO/IEC 23091-2
* VIDFRNG, 1 byte, Value of the VideoFullRangeFlag specified in Rec. ITU-T H.273 | ISO/IEC 23091-2

[ed.: these are inspired from recent JPEG standards that incorporate
H.273 color space parameters]

[ed.: The MATOEFFS parameter is not included because PNG does not support YCbCr]

> This specification uses the existing iCCP chunk to unambiguously signal the color system of an image that uses the Reference PQ EOTF specified in [BT2100-1]. It also allows graceful processing by decoders that do not conform to this specification by recommending fallback values for the gAMA chunk, cHRM chunk, and embedded ICC profile.

Reviewer 3: ```suggestion
When the cICP chunk is present, PNG decoders that recognize it shall ignore the following chunks:
- iCCP
- gAMA 
- cHRM 
- sRGB 
```

[H.273](https://www.itu.int/rec/T-REC-H.273/en) specifies a controlled vocabulary for the parameterization of
color space information:
* Color Primaries
* Transfer Function
* Matrix Coefficients (only required if essence is stored as Y’CbCr)

Use [ITU-T Series H Supplement 19](https://www.itu.int/rec/T-REC-H.Sup19-201910-I) “recommendations” to signal the controlled vocabulary for image formats commonly used for baseband linear broadcasts or file-based Video-on-Demand(VOD) services.
* Tables within this document summarize values most commonly used in broadcast for 47 different standards documents including H.273

## A. References

> ### A.1 Normative references

Reviewer 3: ```suggestion
```

## Basic requirements
* does not break current implementations
* extensible signaling of color space based on H.273
* does not require the presence of iCCP chunk and embedded ICC profiles

## Strawman approach
[H.273](https://www.itu.int/rec/T-REC-H.273/en) specifies a controlled vocabulary for the parameterization of
color space information.

Define a cICP chunk that contains the 7 bytes necessary to carry the
H.273 color space parameters:

* COLPRIMS, 2 bytes, One of the ColourPrimaries enumerated values specified in Rec. ITU-T H.273 | [ISO/IEC 23091-2]
* TRANSFC, 2 bytes, One of the TransferCharacteristics enumerated values specified in Rec. ITU-T H.273 | [ISO/IEC 23091-2]
* VIDFRNG, 1 byte, Value of the VideoFullRangeFlag specified in Rec. ITU-T H.273 | [ISO/IEC 23091-2]

[ed.: these are inspired from recent JPEG standards that incorporate
H.273 color space parameters]

[ed.: The MATOEFFS parameter is not included because PNG does not support YCbCr]

The cICP chunk comes before IDAT chunk.

When the cICP chunk is present, PNG decoders that recognize it shall ignore the following chunks:
- iCCP
- gAMA 
- cHRM 
- sRGB 

> Use [ITU-T Series H Supplement 19](https://www.itu.int/rec/T-REC-H.Sup19-201910-I) “recommendations” to signal the controlled vocabulary for image formats commonly used for baseband linear broadcasts or file-based Video-on-Demand(VOD) services.

Reviewer 3: ```suggestion
NOTE: [ITU-T Series H Supplement 19](https://www.itu.int/rec/T-REC-H.Sup19-201910-I) summarize combinations of H.273 parameters corresponding to common baseband linear broadcasts and file-based Video-on-Demand(VOD) services.
```

> * Tables within this document summarize values most commonly used in broadcast for 47 different standards documents including H.273

Reviewer 3: ```suggestion
```

> ## A. References

Reviewer 3: ```suggestion
## A. References

[ITU-T Series H Supplement 19](https://www.itu.int/rec/T-REC-H.Sup19-201910-I). Series H: Audiovisual and multimedia systems - Usage of video signal type code points
```

# Adding support for HDR imagery to the PNG format

> Authors (alphabetical): Chris Blume, Pierre-Anthony Lemieux, Chris Seeger

Reviewer 3: ```suggestion
Editors: Chris Blume, Pierre-Anthony Lemieux, Chris Seeger
```

## Basic requirements
* does not break current implementations
* extensible signaling of color space based on H.273
* does not require the presence of iCCP chunk and embedded ICC profiles

## Strawman approach
[H.273](https://www.itu.int/rec/T-REC-H.273/en) specifies a controlled vocabulary for the parameterization of
color space information.

Define a cICP chunk that contains the 7 bytes necessary to carry the
H.273 color space parameters:

* COLPRIMS, 2 bytes, One of the ColourPrimaries enumerated values specified in Rec. ITU-T H.273 | [ISO/IEC 23091-2]
* TRANSFC, 2 bytes, One of the TransferCharacteristics enumerated values specified in Rec. ITU-T H.273 | [ISO/IEC 23091-2]
* VIDFRNG, 1 byte, Value of the VideoFullRangeFlag specified in Rec. ITU-T H.273 | [ISO/IEC 23091-2]

[ed.: these are inspired from recent JPEG standards that incorporate
H.273 color space parameters]

[ed.: The MATOEFFS parameter is not included because PNG does not support YCbCr]

The cICP chunk comes before IDAT chunk.

When the cICP chunk is present, PNG decoders that recognize it shall ignore the following chunks:

> - iCCP

Reviewer 2: See above - this is incorrect in the presence of an ICC profile containing an CICP tag.
